pkg/bucket/bucket_controller.go

```diff
@@ -0,0 +1,196 @@
+/* Copyright 2021 The Kubernetes Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package bucket
+
+import (
+	"context"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilversion "k8s.io/apimachinery/pkg/util/version"
+	kube "k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+
+	"sigs.k8s.io/container-object-storage-interface-api/apis/objectstorage.k8s.io/v1alpha1"
+	buckets "sigs.k8s.io/container-object-storage-interface-api/clientset"
+	bucketapi "sigs.k8s.io/container-object-storage-interface-api/clientset/typed/objectstorage.k8s.io/v1alpha1"
+	cosi "sigs.k8s.io/container-object-storage-interface-spec"
+
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// BucketListener manages Bucket objects
+type BucketListener struct {
+	provisionerClient cosi.ProvisionerClient
+	provisionerName   string
+
+	kubeClient   kube.Interface
+	bucketClient buckets.Interface
+	kubeVersion  *utilversion.Version
+}
+
+// NewBucketListener returns a resource handler for Bucket objects
+func NewBucketListener(provisionerName string, client cosi.ProvisionerClient) *BucketListener {
+	bl := &BucketListener{
+		provisionerName:   provisionerName,
+		provisionerClient: client,
+	}
+
+	return bl
+}
+
+// Add attempts to create a bucket for a given bucket. This function must be idempotent
+// Return values
+//    nil - Bucket successfully provisioned
+//    non-nil err - Internal error                                [requeue'd with exponential backoff]
+func (b *BucketListener) Add(ctx context.Context, bucket *v1alpha1.Bucket) error {
```

Comment:
You should create a copy of `bucket`, since you modify [below](https://github.com/kubernetes-sigs/container-object-storage-interface-provisioner-sidecar/pull/49/files#diff-3a83d706de196d2c768a33c079f065a6ca2ad5f62e75d18525d37df05563fcbbR106).

Reply:
yeah already done